server/src/main/java/io/crate/replication/logical/MetadataTracker.java

```diff
@@ -201,36 +214,94 @@ static ClusterState updateMetadata(String subscriptionName, ClusterState subscri
-                    }
+                var updatedIndexMetadataBuilder = IndexMetadata.builder(subscriberIndexMetadata);
+                var updatedMapping = updateIndexMetadataMappings(publisherIndexMetadata, subscriberIndexMetadata);
+                if (updatedMapping != null) {
+                    updatedIndexMetadataBuilder.putMapping(updatedMapping).mappingVersion(publisherIndexMetadata.getMappingVersion());
+                }
+                var updatedSettings = updateIndexMetadataSettings(publisherIndexMetadata, subscriberIndexMetadata, indexScopedSettings);
+                if (updatedSettings != null) {
+                    updatedIndexMetadataBuilder.settings(updatedSettings).settingsVersion(publisherIndexMetadata.getSettingsVersion());
+                }
+                if (updatedMapping != null || updatedSettings != null) {
+                    updatedMetadataBuilder.put(updatedIndexMetadataBuilder.build(), true);
+                    updateClusterState = true;
                 }
             }
         }
-        if (isUpdated) {
+        if (updateClusterState) {
             if (LOGGER.isDebugEnabled()) {
                 LOGGER.debug("Updated index metadata for subscription {}", subscriptionName);
             }
-            return ClusterState.builder(subscriberClusterState).metadata(metadataBuilder).build();
+            return ClusterState.builder(subscriberClusterState).metadata(updatedMetadataBuilder).build();
         } else {
             return subscriberClusterState;
         }
     }
 
+    @Nullable
+    private static MappingMetadata updateIndexMetadataMappings(IndexMetadata publisherIndexMetadata,
+                                                               IndexMetadata subscriberIndexMetadata) {
+        var publisherMapping = publisherIndexMetadata.mapping();
+        var subscriberMapping = subscriberIndexMetadata.mapping();
+        if (publisherMapping != null && subscriberMapping != null) {
+            if (publisherIndexMetadata.getMappingVersion() > subscriberIndexMetadata.getMappingVersion()) {
+                if (LOGGER.isDebugEnabled()) {
+                    LOGGER.debug("Updated index mapping {} for subscription {}", subscriberIndexMetadata.getIndex().getName(), publisherMapping.toString());
+                }
+                return publisherMapping;
+            }
+        }
+        return null;
+    }
+
+    @Nullable
+    private static Settings updateIndexMetadataSettings(IndexMetadata publisherMetadata,
+                                                       IndexMetadata subscriberMetadata,
+                                                       IndexScopedSettings indexScopedSettings) {
+        if (publisherMetadata.getSettingsVersion() < subscriberMetadata.getSettingsVersion()) {
+            return null;
+        }
+        var publisherSettings = publisherMetadata.getSettings().filter(key -> isReplicatableSetting(key, indexScopedSettings));
```

Comment:
As this already loops through the settings, can't we do the logic of the 2nd loop here as well already? would save another loop.

Reply:
done!